src/v/cloud_storage/remote_partition.cc

```diff
@@ -589,12 +589,14 @@ ss::future<> remote_partition::stop() {
         pair.second->unlink();
     }
 
-    for (auto& s : _segments) {
+    // Swap segment map into a local variable to avoid map mutation while the
+    // segments are being stopped.
+    decltype(_segments) segments_to_stop;
+    segments_to_stop.swap(_segments);
```

Comment:
nit maybe this would avoid the decltype

```c++
auto segments_to_stop = std::exchange(_segments, {});
```

Reply:
it seems to compile to a similar instruction sequence, so no penalty in going with std::exchange https://gcc.godbolt.org/z/E66zx7deM (but don't try -Os, it's weird)

Reply:
the benefit of std::exchange is that it's idiomatic compared with decltype--this isn't really generic code.